Time-Dependent-Coupled-Cluster/td-ccsd.py

```diff
@@ -0,0 +1,798 @@
+
+    ### Build RHS side of t2 equations
+    rhs_T2 = MO[o, o, v, v].copy()
+
+    # P_(ab) t_ijae (F_be - 0.5 t_mb F_me)
+    tmp = Fae - 0.5 * np.einsum('mb,me->be', t1, Fme)
+    Pab = np.einsum('ijae,be->ijab', t2, tmp)
+    rhs_T2 += Pab
+    rhs_T2 -= Pab.swapaxes(2, 3)
+
+    # P_(ij) t_imab (F_mj + 0.5 t_je F_me)
+    tmp = Fmi + 0.5 * np.einsum('je,me->mj', t1, Fme)
+    Pij = np.einsum('imab,mj->ijab', t2, tmp)
+    rhs_T2 -= Pij
+    rhs_T2 += Pij.swapaxes(0, 1)
+
+    tmp_tau = build_tau(t1, t2)
+    rhs_T2 += 0.5 * np.einsum('mnab,mnij->ijab', tmp_tau, Wmnij)
+    rhs_T2 += 0.5 * np.einsum('ijef,abef->ijab', tmp_tau, Wabef)
+
+    # P_(ij) * P_(ab)
+    # (ij - ji) * (ab - ba)
+    # ijab - ijba -jiab + jiba
+    tmp = np.einsum('ie,ma,mbej->ijab', t1, t1, MO[o, v, v, o])
+    Pijab = np.einsum('imae,mbej->ijab', t2, Wmbej)
+    Pijab -= tmp
+
+    rhs_T2 += Pijab
+    rhs_T2 -= Pijab.swapaxes(2, 3)
+    rhs_T2 -= Pijab.swapaxes(0, 1)
+    rhs_T2 += Pijab.swapaxes(0, 1).swapaxes(2, 3)
+
+    Pij = np.einsum('ie,abej->ijab', t1, MO[v, v, v, o])
+    rhs_T2 += Pij
+    rhs_T2 -= Pij.swapaxes(0, 1)
+
+    Pab = np.einsum('ma,mbij->ijab', t1, MO[o, v, o, o])
+    rhs_T2 -= Pab
+    rhs_T2 += Pab.swapaxes(2, 3)
+
+    ### Update t1 and t2 amplitudes
+    t1 = rhs_T1 / Dia
+    t2 = rhs_T2 / Dijab
+
+    ### Compute CCSD correlation energy
+    CCSDcorr_E = np.einsum('ia,ia->', F[o, v], t1)
+    CCSDcorr_E += 0.25 * np.einsum('ijab,ijab->', MO[o, o, v, v], t2)
+    CCSDcorr_E += 0.5 * np.einsum('ijab,ia,jb->', MO[o, o, v, v], t1, t1)
+
+    ### Print CCSD correlation energy
+    print('CCSD Iteration %3d: CCSD correlation = %3.12f  '\
+          'dE = %3.5E' % (CCSD_iter, CCSDcorr_E, (CCSDcorr_E - CCSDcorr_E_old)))
+    if (abs(CCSDcorr_E - CCSDcorr_E_old) < E_conv):
+        break
+
+    CCSDcorr_E_old = CCSDcorr_E
+
+print('CCSD iterations took %.2f seconds.\n' % (time.time() - ccsd_tstart))
+
+CCSD_E = SCF_E + CCSDcorr_E
+
+print('\nFinal CCSD correlation energy:     % 16.10f' % CCSDcorr_E)
+print('Total CCSD energy:                 % 16.10f' % CCSD_E)
+if compare_psi4:
+    psi4.driver.p4util.compare_values(psi4.energy('CCSD'), CCSD_E, 6, 'CCSD Energy')
+
+if print_amps:
+    # [::4] take every 4th, [-5:] take last 5, [::-1] reverse order
+    t2_args = np.abs(t2).ravel().argsort()[::2][-5:][::-1]
+    t1_args = np.abs(t1).ravel().argsort()[::4][-5:][::-1]
+
+    print('\nLargest t1 amplitudes')
+    for pos in t1_args:
+        value = t1.flat[pos]
+        inds = np.unravel_index(pos, t1.shape)
+        print('%4d  %4d |   % 5.10f' % (inds[0], inds[1], value))
+
+    print('\nLargest t2 amplitudes')
+    for pos in t2_args:
+        value = t2.flat[pos]
+        inds = np.unravel_index(pos, t2.shape)
+
+
+### TD-CCSD code starts here!!!
+
+#Step 1: Build one and two-particle elements of the similarity-transformed Hamiltonian
+#Equations from J. Gauss and J.F. Stanton, JCP, 103, 9, 1995. Table III (b)
```

Comment:
I wonder, would we be able to host papers like this one that contain equations that we reference specifically? So that readers don't have to go out looking for them, they'd just be available?  Could be a main-repo level directory which contains copies of any papers which get referenced.